### PR TITLE
Support Lambda Platform Metrics

### LINKED ISSUE #108: Support for Metrics
We should decide on a strategy for reporting metrics (such as memory usage, total available memory, billing duration/init duration, etc.).  This may mean making the agent's metrics work inside a lambda function, it may mean using [the Lambda Logs API](https://docs.aws.amazon.com/lambda/latest/dg/runtimes-logs-api.html) to look for and report metrics, it may means leveraging [metric beat](https://www.elastic.co/beats/metricbeat) or [function beat](https://www.elastic.co/beats/functionbeat), or may mean some other thing.

- [ ] Decide on Metrics Strategy 
- [ ] Decide which Metrics to Collect
- [ ] Expose Lambda specific metrics via UI

### LINKED ISSUE #109: Support for Log Forwarding?
We should decide on a strategy for forwarding logs related to AWS Lambda functions.  This may mean leveraging [the AWS Lambda Logs API](https://docs.aws.amazon.com/lambda/latest/dg/runtimes-logs-api.html), may mean leverage [some other elastic product], or may mean [some third thing]

- [ ] Decide on Strategy for Forwarding Logs

### PR DESCRIPTION
**Pull request by jlvoiseux**
### Motivation / Summary
This PR is the continuation of #127, which had become obsolete and non-digestible due to the many changes brought to the extension since the initialization of the PR. The current PR implements the same features, but was rebased to enable a commit-by-commit review approach. It resolves #108.

The general flow is as follows:
- The extension gets started (cold start)
- A metadata container is initialized. This container is populated with the metadata extracted from the first APM payload sent by the agent
- The log processing goroutine now handles a new type of `SubEventType` alongside `RuntimeDone`: `Report`. This report contains the metrics that we want to send to the APM Server. These metrics are packaged inside an APM payload (alongside the extracted metadata) in a dedicated function.

### Files / Added Elements
- 944e71ffca910c8951dfa07b1b507520407edf53 : Add a new package `model` which is a partial reproduction of the data model used by the [Go APM Agent](https://github.com/elastic/apm-agent-go/blob/main/model/model.go)
- 1ec30ab1bbf383949b8084bfa5a1ae0089f7747d : Add a version utility file that is used to populate metadata
- 10e2c77dc8b3af3325fc58058be22c86c1d7ff66 and 3ec3bf9f4c9968c42113593ff91a484f0a4b9e29 : Functions that extract metadata from a given APM payload and populate a dedicated container. Since metadata extraction is tied to APM data, it is part of the `extension` package. Metadata is only extracted from the first APM payload received during the current lifetime of the extension.
- a515a6d456b3600bd1cc8c288c183b82ae2f88e3 and c2b056a07d6666ac300fa8303d1945c9cbae5700 : Functions that turn the platform report provided by the Lambda runtime into a payload to be sent to the APM Server. Since metrics extraction is tied to logs data, it is part of the `logsapi` package.
- 2b651c8d239bc930125bbaf6653336c1406035f2 and 614be84d8f0930c36d73a6d813fb8f745eab7ace : `main` updates to modify the goroutines in order to extract metadata and process metrics.

### Limitations / Elements to discuss
- After several tests with [two](https://github.com/aws-samples/aws-lambda-extensions/tree/main/go-example-logs-api-extension) [sample](https://github.com/aws-samples/aws-lambda-extensions/tree/main/s3-logs-extension-demo-zip-archive) extensions with two different logs processing philosophies, we can [conclude](https://github.com/aws-samples/aws-lambda-extensions/blob/949b01bd789f90bd39cc718fc11935c4109dfcf7/s3-logs-extension-demo-zip-archive/extensionssrc/extensions/logs_api_http_extension.py#L29) that the metrics report corresponding to the invocation `i` will be received during the execution of the invocation `i+1`.
- The current implementation changes the Agent name to `apm-lambda-extension` when the metadata is attached to the generated metrics. While this makes sense in the framework of observability, the current APM UI is not equipped accordingly, and the service is associated to an unknown runtime:
image
My suggestion would be to stop altering the metadata for now to provide a consistent experience to end users.
- The current package architecture can also be discussed. I wished I was able to keep `logsapi` and `extension` completely separate, but in the current implementation, `logsapi.ProcessLogs()` takes an `apmServerTransport` in argument in order to enqueue APM metrics payloads.
- Several `faas` fields will be added to the `metricset` object in the APM data model (PR to be created). The subsequent modifications are beyond the scope of this initial PR, which aims to provide an initial implementation of Lambda metrics generation. 

### How to test
- Clone the extension repo
- Follow the instructions presented in the [following](https://github.com/jlvoiseux/elastic-apm-lambda-dev-sample) sample repo
- After executing your sample Lambda function a few times, you should see documents corresponding to your Lambda Platform report in your Elastic Cloud instance:
```
// Lambda Report
REPORT RequestId: 6862dc9b-3266-470a-9979-f39c21164f8c	Duration: 94.65 ms	Billed Duration: 95 ms	Memory Size: 128 MB	Max Memory Used: 77 MB	Init Duration: 479.30 ms	
```
image
image

### CODE REVIEW

File: apm-lambda-extension/logsapi/process_metrics_test.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+	}
+	mc.Metadata.Service = model.Service{
+		Name:      os.Getenv("AWS_LAMBDA_FUNCTION_NAME"),
+		Agent:     model.Agent{Name: "python", Version: "6.7.2"},
+		Language:  model.Language{Name: "python", Version: "3.9.8"},
+		Runtime:   model.Runtime{Name: os.Getenv("AWS_EXECUTION_ENV")},
+		Framework: model.Framework{Name: "AWS Lambda"},
+	}
+	mc.Metadata.Process = model.Process{}
+	mc.Metadata.System = model.System{}
```
Comment (axw) [inline apm-lambda-extension/logsapi/process_metrics_test.go:84]:
```suggestion
```

Aren't they already zero valued?

**Review by axw — CHANGES_REQUESTED** [review]
Generally looks good, mostly minor comments.

File: apm-lambda-extension/logsapi/process_metrics.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+	if metadataContainer.Metadata != nil {
+		//TODO : Discuss relevance of displaying extension name
+		metadataContainer.Metadata.Service.Agent.Name = "apm-lambda-extension"
+		metadataContainer.Metadata.Service.Agent.Version = extension.Version
+		metadataJson, err := json.Marshal(metadataContainer)
+		if err != nil {
+			return extension.AgentData{}, err
+		}
+		metricsData = append(metadataJson, []byte("\n")...)
+	}
```
[the remark this answers is not in the document]
Comment (felixbarny) [reply apm-lambda-extension/logsapi/process_metrics.go:107]: > we should probably defer modifying the agent name for now anyway

that makes sense to me

File: apm-lambda-extension/extension/process_metadata_test.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+		if err := ProcessMetadata(agentData, &mc); err != nil {
+			t.Fail()
+		}
+	}
+
+	if mc.Metadata == nil {
+		if err := ProcessMetadata(badAgentData, &mc); err != nil {
+			t.Fail()
+		}
+	}
```
[the remark this answers is not in the document]
Comment (jlvoiseux) [reply apm-lambda-extension/extension/process_metadata_test.go]: Fixed in 11e874b

File: apm-lambda-extension/logsapi/process_metrics_test.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+	"context"
+	"fmt"
+	"os"
+	"testing"
+	"time"
+
+	"elastic/apm-lambda-extension/extension"
+	"elastic/apm-lambda-extension/model"
+
+	"gotest.tools/assert"
```
[the remark this answers is not in the document]
Comment (jlvoiseux) [reply apm-lambda-extension/logsapi/process_metrics_test.go]: Indeed, great catch! Fixed in 11e874b

File: apm-lambda-extension/logsapi/process_metrics_test.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+	rawBytes, err := ProcessPlatformReport(context.Background(), &mc, &event, logEvent)
+	if err != nil {
+		t.Fail()
+	}
+	requestBytes, err := extension.GetUncompressedBytes(rawBytes.Data, "")
+	if err != nil {
+		t.Fail()
+	}
+	assert.Equal(t, string(requestBytes), desiredOutput)
```
[the remark this answers is not in the document]
Comment (jlvoiseux) [reply apm-lambda-extension/logsapi/process_metrics_test.go]: Fixed in 11e874b

File: apm-lambda-extension/logsapi/process_metrics_test.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+	mc.Metadata.System = model.System{}
+
+	rawBytes, err := ProcessPlatformReport(context.Background(), &mc, &event, logEvent)
+	if err != nil {
+		t.Fail()
+	}
+	requestBytes, err := extension.GetUncompressedBytes(rawBytes.Data, "")
+	if err != nil {
+		t.Fail()
+	}
```
[the remark this answers is not in the document]
Comment (jlvoiseux) [reply apm-lambda-extension/logsapi/process_metrics_test.go]: Fixed in 11e874b

File: apm-lambda-extension/logsapi/process_metrics_test.go (hunk 1 above)
Comment (jlvoiseux) [reply apm-lambda-extension/logsapi/process_metrics_test.go]: Fixed in 11e874b

File: apm-lambda-extension/logsapi/process_metrics.go [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+		metricsContainer.Metrics.Labels["faas.coldstart"] = false
+	}
+	// System
+	metricsContainer.Add("system.memory.total", nil, float64(platformReportMetrics.MemorySizeMB)*convMB2Bytes)                                             // Unit : Bytes
+	metricsContainer.Add("system.memory.actual.free", nil, float64(platformReportMetrics.MemorySizeMB-platformReportMetrics.MaxMemoryUsedMB)*convMB2Bytes) // Unit : Bytes
+
+	// Raw Metrics
+	// AWS uses binary multiples to compute memory : https://aws.amazon.com/about-aws/whats-new/2020/12/aws-lambda-supports-10gb-memory-6-vcpu-cores-lambda-functions/
+	metricsContainer.Add("aws.lambda.metrics.TotalMemory", nil, float64(platformReportMetrics.MemorySizeMB)*convMB2Bytes)   // Unit : Bytes
+	metricsContainer.Add("aws.lambda.metrics.UsedMemory", nil, float64(platformReportMetrics.MaxMemoryUsedMB)*convMB2Bytes) // Unit : Bytes
```
[the remark this answers is not in the document]
Comment (jlvoiseux) [reply apm-lambda-extension/logsapi/process_metrics.go]: Fixed in c018cb6

File: apm-lambda-extension/logsapi/process_metrics.go [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+	InitDurationMs   float32 `json:"initDurationMs"`
+}
+
+type MetricsContainer struct {
+	Metrics *model.Metrics `json:"metricset"`
+}
+
+// Add adds a metric with the given name, labels, and value,
+// The labels are expected to be sorted lexicographically.
+func (mc MetricsContainer) Add(name string, labels []model.MetricLabel, value float64) {
```
[the remark this answers is not in the document]
Comment (jlvoiseux) [reply apm-lambda-extension/logsapi/process_metrics.go]: Fixed in c018cb6

File: apm-lambda-extension/logsapi/process_metrics.go [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+	// In AWS Lambda, the Timeout is configured as an integer number of seconds. We use this assumption to derive the Timeout from
+	// - The epoch corresponding to the end of the current invocation (its "deadline")
+	// - The epoch corresponding to the start of the current invocation
+	// - The multiplication / division then rounds the value to obtain a number of ms that can be expressed a multiple of 1000 (see initial assumption)
+	metricsContainer.Add("aws.lambda.metrics.Timeout", nil, math.Ceil(float64(functionData.DeadlineMs-functionData.Timestamp.UnixMilli())/1e3)*1e3) // Unit : Milliseconds
+
+	metricsJson, err := json.Marshal(metricsContainer)
+	if err != nil {
+		return extension.AgentData{}, err
+	}
```
[the remark this answers is not in the document]
Comment (jlvoiseux) [reply apm-lambda-extension/logsapi/process_metrics.go]: Fixed in c018cb6

File: apm-lambda-extension/logsapi/process_metrics.go [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+	metricsContainer.Add("aws.lambda.metrics.UsedMemory", nil, float64(platformReportMetrics.MaxMemoryUsedMB)*convMB2Bytes) // Unit : Bytes
+	metricsContainer.Add("aws.lambda.metrics.Duration", nil, float64(platformReportMetrics.DurationMs))                     // Unit : Milliseconds
+	metricsContainer.Add("aws.lambda.metrics.BilledDuration", nil, float64(platformReportMetrics.BilledDurationMs))         // Unit : Milliseconds
+	metricsContainer.Add("aws.lambda.metrics.ColdStartDuration", nil, float64(platformReportMetrics.InitDurationMs))        // Unit : Milliseconds
+	// In AWS Lambda, the Timeout is configured as an integer number of seconds. We use this assumption to derive the Timeout from
+	// - The epoch corresponding to the end of the current invocation (its "deadline")
+	// - The epoch corresponding to the start of the current invocation
+	// - The multiplication / division then rounds the value to obtain a number of ms that can be expressed a multiple of 1000 (see initial assumption)
+	metricsContainer.Add("aws.lambda.metrics.Timeout", nil, math.Ceil(float64(functionData.DeadlineMs-functionData.Timestamp.UnixMilli())/1e3)*1e3) // Unit : Milliseconds
```
[the remark this answers is not in the document]
Comment (jlvoiseux) [reply apm-lambda-extension/logsapi/process_metrics.go]: While the Timeout was not part of the initial specification, I believe that sending in to ES (and Kibana) is useful for visualisation purpose. If they have access to this metric, end-users can build a dashboard that would help them decide whether or not the timeout they specified for their Lambda functions is adequate, given execution time measurements.

File: apm-lambda-extension/logsapi/process_metrics.go [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+	metricsContainer.Add("system.memory.actual.free", nil, float64(platformReportMetrics.MemorySizeMB-platformReportMetrics.MaxMemoryUsedMB)*convMB2Bytes) // Unit : Bytes
+
+	// Raw Metrics
+	// AWS uses binary multiples to compute memory : https://aws.amazon.com/about-aws/whats-new/2020/12/aws-lambda-supports-10gb-memory-6-vcpu-cores-lambda-functions/
+	metricsContainer.Add("aws.lambda.metrics.TotalMemory", nil, float64(platformReportMetrics.MemorySizeMB)*convMB2Bytes)   // Unit : Bytes
+	metricsContainer.Add("aws.lambda.metrics.UsedMemory", nil, float64(platformReportMetrics.MaxMemoryUsedMB)*convMB2Bytes) // Unit : Bytes
+	metricsContainer.Add("aws.lambda.metrics.Duration", nil, float64(platformReportMetrics.DurationMs))                     // Unit : Milliseconds
+	metricsContainer.Add("aws.lambda.metrics.BilledDuration", nil, float64(platformReportMetrics.BilledDurationMs))         // Unit : Milliseconds
+	metricsContainer.Add("aws.lambda.metrics.ColdStartDuration", nil, float64(platformReportMetrics.InitDurationMs))        // Unit : Milliseconds
```
[the remark this answers is not in the document]
Comment (jlvoiseux) [reply apm-lambda-extension/logsapi/process_metrics.go]: Corrected in d5952d6

File: apm-lambda-extension/extension/process_metadata.go [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+	Metadata *model.Metadata `json:"metadata"`
+}
+
+func ProcessMetadata(data AgentData, container *MetadataContainer) error {
+	uncompressedData, err := GetUncompressedBytes(data.Data, data.ContentEncoding)
+	if err != nil {
+		return errors.New(fmt.Sprintf("Error uncompressing agent data for metadata extraction : %v", err))
+	}
+	decoder := json.NewDecoder(bytes.NewReader(uncompressedData))
+	for {
```
[the remark this answers is not in the document]
Comment (jlvoiseux) [reply apm-lambda-extension/extension/process_metadata.go]: The loop was replaced by a single scanner iteration in 62501a6

File: apm-lambda-extension/go.mod [hunk 13]
```diff
@@ -15,9 +15,15 @@ require (
 	gotest.tools v2.2.0+incompatible
 )
 
+require (
+	go.elastic.co/apm v1.15.0
```
Comment (axw) [inline apm-lambda-extension/go.mod:19]: We should use `go.elastic.co/apm/v2`, the current major version.

File: apm-lambda-extension/logsapi/process_metrics.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (axw) [reply apm-lambda-extension/logsapi/process_metrics.go]: Sounds good, thanks.

**Review by axw — APPROVED** [review]
LGTM, just one thing (we should use Go agent v2).

I assume we'll want the labels to be recorded as `faas.*`, rather than `labels.faas_*`, right? Is your plan that once the server is updated, we'll come back and update the Go agent's model and the extension to send them as metricset fields? Similar to what we do for transaction and span properties: https://github.com/elastic/apm-agent-go/blob/94d06659448a555142c087721a79fff3ee732b49/model/model.go#L798-L804

**Comment by jlvoiseux** [comment]
@axw Thank you for the review. I have updated the model to `v2`.
You are absolutely right regarding the `faas` labels. There was nothing preventing us from modifying the APM model in parallel to this PR, but given the number of changes introduced, I thought that the added complexity of having to test the PR alongside a modified APM Model was going to be too much at once. However, that is definitely the next step once we merge this.

**Pull request merged**